### PR TITLE
 Add deprecation warnings for usage of `ALL`

### LINKED ISSUE #316: Add deprecation notices to `ALL`
**Is your feature request related to a problem? Please describe.**
To enforce consumers to use `RAPIDS` add deprecation notices when a project uses the older magic values.

**Describe the solution you'd like**
Large CMake deprecation warning that this will be released in the N+1 version of rapids-cmake with documentation on how to fix. 
Would use `message(AUTHOR_WARNING`

### PR DESCRIPTION
**Pull request by robertmaynard**
## Description
Adds deprecation notices to usage of `ALL` with `CMAKE_CUDA_ARCHITECTURES` and `rapids_cmake_set_architectures`. This is done to remove the ambiguity between `all` ( all archs ) and `ALL` ( all archs that RAPIDS supports ). Instead projects should use `RAPIDS` as the magic value.

Fixes #316 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/rapids-cmake/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The `cmake-format.json` is up to date with these changes.
- [x] I have added new files under rapids-cmake/
   - [x] I have added include guards (`include_guard(GLOBAL)`)
   - [x] I have added the associated docs/ rst file and update the api.rst

### CODE REVIEW

File: rapids-cmake/cuda/detail/architectures_policy.cmake [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  list(APPEND CMAKE_MESSAGE_CONTEXT "rapids.cuda.architectures_policy")
+
+  include("${rapids-cmake-dir}/cmake/detail/policy.cmake")
+
+  set(value ${${mode_variable}})
+  set(new_value ${value})
+  if(value STREQUAL "ALL")
+    set(new_value "RAPIDS")
+    if(called_from STREQUAL "FROM_INIT")
+      rapids_cmake_policy(DEPRECATED_IN 23.02 REMOVED_IN 23.06 MESSAGE
```
Comment (bdice) [inline rapids-cmake/cuda/detail/architectures_policy.cmake:42]: Why wait until 23.06?

File: testing/cuda/CMakeLists.txt [hunk 2]
```diff
@@ -40,3 +40,8 @@ add_cmake_config_test( set_arch-existing.cmake )
 add_cmake_config_test( set_arch-invalid-mode.cmake )
 add_cmake_config_test( set_arch-native.cmake )
 add_cmake_config_test( set_arch-rapids.cmake )
+
+set(deprecateds_message "rapids-cmake policy [deprecated=23.02 removed=23.06]: Usage of `ALL`")
```
Comment (bdice) [inline testing/cuda/CMakeLists.txt:44]: Does the “s” mean something in “deprecateds”?
```suggestion
set(deprecated_message "rapids-cmake policy [deprecated=23.02 removed=23.06]: Usage of `ALL`")
```

**Review by bdice — COMMENTED** [review]
All looks good! Two minor questions.

File: rapids-cmake/cuda/detail/architectures_policy.cmake (hunk 1 above)
Comment (robertmaynard) [reply rapids-cmake/cuda/detail/architectures_policy.cmake:42]: We have non rapids projects  ( cuco / nvbench ) that aren't on the same release candence, so I want to give ample time to allow for all users of rapids-cmake to upgrade.

File: testing/cuda/CMakeLists.txt (hunk 2 above)
Comment (robertmaynard) [reply testing/cuda/CMakeLists.txt]: Corrected

**Comment by robertmaynard** [comment]
With PR's open for all RAPIDS projects to use the `RAPIDS` magic value, I think we are unblocked on merging this.

File: rapids-cmake/cuda/detail/architectures_policy.cmake (hunk 1 above)
Comment (bdice) [reply rapids-cmake/cuda/detail/architectures_policy.cmake]: Aha! Sounds good.

**Pull request merged**